src/bmp/decoder.rs

```diff
@@ -1314,11 +1317,19 @@ impl<'a, R: 'a + Read + Seek> ImageDecoder<'a> for BMPDecoder<R> {
     }
 
     fn into_reader(self) -> ImageResult<Self::Reader> {
-        Ok(BmpReader(Cursor::new(self.read_image()?), PhantomData))
+        if self.total_bytes() > usize::max_value() as u64 {
+            return Err(ImageError::InsufficientMemory);
+        }
+
+        let mut buf = vec![0; self.total_bytes() as usize];
+        self.read_image(&mut buf)?;
+
+        Ok(BmpReader(Cursor::new(buf), PhantomData))
     }
 
-    fn read_image(mut self) -> ImageResult<Vec<u8>> {
-        self.read_image_data()
+    fn read_image(mut self, buf: &mut [u8]) -> ImageResult<()> {
+        assert!(buf.len() as u64 == self.total_bytes());
```

Comment:
This comparison should consider the case that `usize > u64`. One possible alternative assertion:

>  `assert_eq!(u64::try_from(buf.len()), Ok(self.total_bytes());`

Reply:
Instances of similar code appear in other decoders.

Reply:
I think this case is fine because to trigger the overflow the user would have be running on a >64bit machine _and_ somehow actually produce a 16+ exabyte buffer. But your proposed assertion seems clean enough that I'd be fine switching it.